gix-dir/src/walk/readdir.rs

```diff
@@ -59,12 +59,24 @@ pub(super) fn recursive(
         );
         current.push(file_name);
 
+        let disk_kind = {
+            let m = current.metadata().map_err(|err| Error::DirEntryFileType {
```

Comment:
I really tried hard to avoid checking each file on disk, but maybe that's ultimately futile as one can't even assume the disk isn't changed when finding a file in the index. Maybe that's also the source of subtle (and untested) bugs.

Also I hope that in practice, there will be no noticeable performance penalty.

Reply:
I'm definitely happy to restore the ability to skip stat().
It's just that all existing callers in the repo call stat() regardless (not sure, will re-check).
Obviously that doesn't apply to public-facing APIs so I might need to revisit